### PR TITLE
Bug fix non unique address view ids

### PR DESCRIPTION
**Pull request by nauaros**
## Summary
- Fix bug on `FormAddressItem` so each one of the subitems have a unique view identifier.
- Update `CardComponent` tests to use the address filling helper method.
- Add missing translations for the Affirm payment method.
- Update layout issues on the Affirm payment method.
- Align the modal view's title leading edge with the form's leading edge.

### CODE REVIEW

File: Adyen/Utilities/ViewIdentifierBuilder.swift [hunk 1]
```diff
@@ -16,7 +16,9 @@ public enum ViewIdentifierBuilder {
     ///    it serves as a name space for the identifer, to avoid identifier collisions.
     ///   - postfix: This should uniquely identify the view inside the scope provided by the `scopeInstance`.
     public static func build(scopeInstance: Any, postfix: String) -> String {
-        let scope = (scopeInstance as? UIView)?.accessibilityIdentifier ?? String(describing: scopeInstance)
-        return "\(scope).\(postfix)"
+        var scope = (scopeInstance as? UIView)?.accessibilityIdentifier
```
Comment (kaphacius) [inline Adyen/Utilities/ViewIdentifierBuilder.swift:19]: maybe like this?
```swift
    public static func build(scopeInstance: Any, postfix: String) -> String {
        let scope = (scopeInstance as? UIView).flatMap(\.accessibilityIdentifier) ??
            (scopeInstance as? FormItem).flatMap(\.identifier) ??
            String(describing: scopeInstance)
        return "\(scope).\(postfix)"
    }
```

File: AdyenDropIn/Views/ModalToolbar.swift [hunk 2]
```diff
@@ -11,7 +11,7 @@ internal final class ModalToolbar: UIView, AnyNavigationBar {
     private let style: NavigationStyle
     internal var onCancelHandler: (() -> Void)?
     private let title: String?
-    private let paddingWithMarginCorrection: CGFloat = 20
+    private let paddingWithMarginCorrection: CGFloat = 16
```
Comment (kaphacius) [inline AdyenDropIn/Views/ModalToolbar.swift:14]: why this change?

File: Tests/AdyenTests/Card Tests/CardComponentTests.swift [hunk 3]
```diff
@@ -785,47 +792,43 @@ class CardComponentTests: XCTestCase {
...
+        // When
+        wait(for: .seconds(1))
 
-            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(1)) {
-                XCTAssertTrue(houseNumberItemView!.alertLabel.isHidden)
-                XCTAssertFalse(addressItemView!.alertLabel.isHidden)
-                XCTAssertFalse(cityItemView!.alertLabel.isHidden)
-                XCTAssertFalse(postalCodeItemView!.alertLabel.isHidden)
+        let view: UIView = sut.viewController.view
+
```
Comment (kaphacius) [inline Tests/AdyenTests/Card Tests/CardComponentTests.swift:799]: `// Then` 😉 
or in line 808?

File: Adyen/Assets/ru-RU.lproj/Localizable.strings [hunk 4]
```diff
@@ -108,9 +108,11 @@
...
 "adyen.giftcard.noBalance" = "На этой подарочной карте нет средств";
 "adyen.partialPayment.remainingBalance" = "Остаток на балансе составит %@";
+"adyen.partialPayment.payRemainingAmount" = "Выбрать другой способ оплаты оставшейся суммы %@";
 "adyen.card.taxNumber.label" = "Дата рождения владельца карты (ГГММДД) или регистрационный номер предприятия (10 цифр)";
 "adyen.card.taxNumber.placeholder" = "ГГММДД / 0123456789";
 "adyen.card.taxNumber.invalid" = "Неверная дата рождения владельца карты или регистрационного номера предприятия";
 "adyen.card.encryptedPassword.label" = "Первые 2 цифры пароля карты";
 "adyen.card.encryptedPassword.placeholder" = "";
 "adyen.card.encryptedPassword.invalid" = "Неверный пароль";
+"adyen.affirm.deliveryAddressToggle.title" = "Доставить по другому адрес";
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Assets/ru-RU.lproj/Localizable.strings:118]: 😆

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift [hunk 5]
```diff
@@ -166,9 +177,32 @@ public final class FormAddressItem: FormValueItem<PostalAddress, AddressStyle>,
...
+    }
+    
+    private func updateItems(identifier: String?) {
+        let items = items + [headerItem, countrySelectItem]
+        for item in items {
+            if let item = item as? FormSplitItem {
+                update(item: item.leftItem, identifier: identifier)
+                update(item: item.rightItem, identifier: identifier)
+                continue
+            }
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:192]: this can be generic using `FormItem.subitems`:

```Swift
for item in items {
             for subitem in item.subitems {
                  update(item: subitem, identifier: identifier)
             }
             update(item: item, identifier: identifier)
}
```

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift [hunk 6]
```diff
@@ -166,9 +177,32 @@ public final class FormAddressItem: FormValueItem<PostalAddress, AddressStyle>,
...
+            update(item: item, identifier: identifier)
         }
     }
     
+    private func update(item: FormItem, identifier: String?) {
+        let className = String(describing: self)
+        let currentIdentifier = item.identifier
+        let newIdentifier = currentIdentifier?.replacingOccurrences(of: className, with: identifier ?? "")
+        item.identifier = newIdentifier
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:201]: this algorithm makes an assumption that the currentIdentifier contains the className, wondering if there is a better way 🤔

File: Adyen/UI/Form/Items/Spacer/FormSpacerItemView.swift [hunk 7]
```diff
@@ -29,7 +29,8 @@ public final class FormSpacerItemView: FormItemView<FormSpacerItem> {
 
     private lazy var stackView: UIView = {
         let stackView = UIStackView(arrangedSubviews: arrangedSubviews)
-        stackView.axis = .horizontal
+        // TODO: FIXME
+        stackView.axis = .vertical
```
[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/UI/Form/Items/Spacer/FormSpacerItemView.swift:33]: it is fixed, he just forgot to remove the comment I think, right?

File: Adyen/Utilities/ViewIdentifierBuilder.swift (hunk 1 above)
Comment (mohammedDehairy) [reply Adyen/Utilities/ViewIdentifierBuilder.swift:19]: or :

```Swift
[
(scopeInstance as? UIView).flatMap(\.accessibilityIdentifier),
(scopeInstance as? FormItem).flatMap(\.identifier),
String(describing: scopeInstance)
].compactMap { $0 }.first
```

Comment (kaphacius) [reply Adyen/Utilities/ViewIdentifierBuilder.swift:19]: nice!

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift [hunk 8]
```diff
@@ -166,9 +177,32 @@ public final class FormAddressItem: FormValueItem<PostalAddress, AddressStyle>,
         case .stateOrProvince:
             observers[.stateOrProvince] = bind(item.publisher, to: self, at: \.value.stateOrProvince)
             publisherObservers[.stateOrProvince] = observe(publisher) { item.value = $0.stateOrProvince ?? "" }
+        case .country:
+            break
+        }
+    }
+    
+    private func updateItems(identifier: String?) {
```
[the remark this answers is not in the document]
Comment (kaphacius) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:185]: looks like there is a simpler way
```swift
    private func updateItems(identifier: String?) {
        let updateAction: (FormItem) -> Void = { self.update(item: $0, identifier: identifier ?? "") }
        (items.flatMap(\.flatSubitems) + [headerItem, countrySelectItem])
            .forEach(updateAction)
    }
```

File: Adyen/Assets/ja-JP.lproj/Localizable.strings [hunk 9]
```diff
@@ -31,7 +31,7 @@
 "adyen.card.stored.expires" = "%@ に期限が切れます";
 "adyen.dropIn.stored.title" = "%@のお支払いをご確認下さい";
 "adyen.dropIn.preselected.openAll.title" = "お支払方法を変更する";
-"adyen.continueTo" = "次へ進む：%@";
+"adyen.continueTo" = "次へ進む： %@";
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/Assets/ja-JP.lproj/Localizable.strings]: fixed

File: Adyen/Assets/ja-JP.lproj/Localizable.strings [hunk 10]
```diff
@@ -47,7 +47,7 @@
 "adyen.blik.code" = "6桁のコード";
 "adyen.blik.help" = "バンキングアプリからコードを取得してください。";
 "adyen.blik.placeholder" = "123–456";
-"adyen.preauthorizeWith" = "次で事前認証する：%@";
+"adyen.preauthorizeWith" = "次で事前認証する： %@";
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/Assets/ja-JP.lproj/Localizable.strings]: fixed

File: Adyen/Core/Components/AbstractPersonalInformationComponent/AddressFormItemInjector.swift [hunk 11]
```diff
@@ -22,11 +22,9 @@ internal final class AddressFormItemInjector: FormItemInjector, Localizable {
 
     /// :nodoc:
     internal lazy var item: FormAddressItem = {
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/Core/Components/AbstractPersonalInformationComponent/AddressFormItemInjector.swift]: Done

File: Adyen/UI/Form/Items/Spacer/FormSpacerItemView.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/UI/Form/Items/Spacer/FormSpacerItemView.swift:32]: Yeah right, already removed the comment. Thanks

File: AdyenDropIn/Views/ModalToolbar.swift (hunk 2 above)
Comment (nauaros) [reply AdyenDropIn/Views/ModalToolbar.swift:14]: Arjen requested a fix on the modal view title margin so it could be aligned with the form's fields

**Comment by mohammedDehairy** [comment]
also, almost all checks are failing, can you please fix them :-)

File: Adyen/Assets/Generated/LocalizationKey.swift [hunk 12]
```diff
@@ -2,7 +2,7 @@
 // Copyright (c) 2021 Adyen N.V.
 //
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
-// This file is autogenerated. Please do not modify it.
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/Assets/Generated/LocalizationKey.swift:5]: SwiftFormat in action

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift [hunk 13]
```diff
@@ -166,9 +177,32 @@ public final class FormAddressItem: FormValueItem<PostalAddress, AddressStyle>,
...
+            if let item = item as? FormSplitItem {
+                update(item: item.leftItem, identifier: identifier)
+                update(item: item.rightItem, identifier: identifier)
+                continue
+            }
+            update(item: item, identifier: identifier)
         }
     }
     
+    private func update(item: FormItem, identifier: String?) {
```
Comment (descorp) [inline Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:197]: Why do we need to go thought every item and check if it identifier contains class name?
You can simply assign identifier to SplitItem and then recalculate identifiers for subitems via ViewIdentifierBuilder

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:185]: Updated by a more functional style one

File: Tests/AdyenTests/Card Tests/CardComponentTests.swift (hunk 3 above)
Comment (nauaros) [reply Tests/AdyenTests/Card Tests/CardComponentTests.swift:799]: Ups

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 5 above)
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift]: Improved

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 6 above)
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:194]: Right now, every item created on `FormFullAddressItem` is using `self` as the scope for the identifier build. I tried passing the identifier as part of the initializer parameters but we ran into a problem. In some scenarios, the address item is created at the parent class (e.g. AbstractPersonalInformationComponent) so we gotta have a way to change it afterwards, that's why I used a property observer on the `identifier` variable.

Since all items have this identifier format: `FormFullAddressItem.xxx`, it was easier to replace `FormFullAddressItem` by the newly assigned identifier of the address item.

If you wanna know more, glad to jump into a call to show you the problem and why I solved it this way.

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 13 above)
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:190]: Check my answer to @mohammedDehairy's comment

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (kaphacius) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:185]: I think it's better to check here if the `identifier` is `nil`, otherwise it will be unwrapped for every call of the `update` method

[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:185]: The logic behind it is that if a nil identifier is passed we simply assign an empty string. Do you think we should be doing something different?

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 6 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:194]: So no way to pre-se identifir?

File: Adyen/Assets/Generated/LocalizationKey.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Assets/Generated/LocalizationKey.swift:5]: you can change and commit without building
Fastest git on the west :D

File: Adyen/UI/Form/Items/Address/FormFullAddressItem.swift (hunk 6 above)
Comment (mohammedDehairy) [reply Adyen/UI/Form/Items/Address/FormFullAddressItem.swift:194]: observer on the `identifier`?, that's too complicated and a bit too much, it should be much simpler than this, lets have a call to understand the problem.

File: AdyenDropIn/Utilities/ComponentManager.swift [hunk 14]
```diff
@@ -50,7 +50,7 @@ internal final class ComponentManager {
 
         // Paid section
         let amountString: String = remainingAmount.map(\.formatted) ??
-            localizedString(.amount, configuration.localizationParameters)
+            localizedString(.partialPaymentPayRemainingAmount, configuration.localizationParameters)
```
Comment (descorp) [inline AdyenDropIn/Utilities/ComponentManager.swift:53]: Why this change? How is it connected to a address identifier?

This also doesn't look correct.
In case there is no `remainingAmount` (that is probably going to be invalid state, but still) you will have footer "Select payment method for the remaining Select payment method for the remaining"  🤔

File: Adyen/Assets/en-US.lproj/Localizable.strings [hunk 15]
```diff
@@ -110,13 +110,12 @@
 "adyen.giftcard.noBalance" = "This gift card has zero balance";
 "adyen.partialPayment.remainingBalance" = "Remaining balance will be %@";
 "adyen.partialPayment.payRemainingAmount" = "Select payment method for the remaining %@";
-"adyen.amount" = "amount";
```
Comment (descorp) [inline Adyen/Assets/en-US.lproj/Localizable.strings:113]: this must remain!

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift [hunk 16]
```diff
@@ -94,15 +94,17 @@ class AffirmComponentTests: XCTestCase {
         sut.deliveryAddressToggleItem.value = true
         
         // Then
-        XCTAssertFalse(sut.deliveryAddressItem.isHidden.wrappedValue)
+        let deliveryAddressItem = try XCTUnwrap(sut.deliveryAddressItem)
```
Comment (mohammedDehairy) [inline Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:97]: Nice!

**Review by mohammedDehairy — DISMISSED** [review]

File: AdyenDropIn/Utilities/ComponentManager.swift (hunk 14 above)
Comment (nauaros) [reply AdyenDropIn/Utilities/ComponentManager.swift:53]: There were some conflicts with develop, amount wasn't inside `LocalizationKey.swift`

Comment (mohammedDehairy) [reply AdyenDropIn/Utilities/ComponentManager.swift:53]: true, this is wrong, this one should be `.amount`

Comment (nauaros) [reply AdyenDropIn/Utilities/ComponentManager.swift]: Fixed

File: Adyen/UI/Form/Items/Address/AddressViewModel.swift [hunk 17]
```diff
@@ -43,7 +44,7 @@ internal struct AddressViewModel {
                                                        .postalCode: .postalCodeFieldPlaceholder,
                                                        .apartment: .apartmentSuiteFieldPlaceholder],
                                          optionalFields: [.apartment],
-                                         schema: AddressField.allCases.map { .item($0) })
+                                         schema: AddressField.allCases.filter { $0 != .country }.map { .item($0) })
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/UI/Form/Items/Address/AddressViewModel.swift:47]: I'd like to keep it this way

**Review by mohammedDehairy — APPROVED** [review]

**Review by descorp — APPROVED** [review]

**Review by mohammedDehairy — APPROVED** [review]
